pypose/module/pf.py

```diff
@@ -0,0 +1,204 @@
+
+        >>> import torch, pypose as pp
+        >>> class NTI(pp.module.System):
+        ...     def __init__(self):
+        ...         super().__init__()
+        ...
+        ...     def state_transition(self, state, input, t=None):
+        ...         return state.cos() + input
+        ...
+        ...     def observation(self, state, input, t):
+        ...         return state.sin() + input
+
+        2. Create a model and filter
+
+        >>> model = NTI()
+        >>> pf = pp.module.PF(model)
+
+        3. Prepare data
+
+        >>> T, N = 5, 2 # steps, state dim
+        >>> states = torch.zeros(T, N)
+        >>> inputs = torch.randn(T, N)
+        >>> observ = torch.zeros(T, N)
+        >>> # std of transition, observation, and estimation
+        >>> q, r, p = 0.1, 0.1, 10
+        >>> Q = torch.eye(N) * q**2
+        >>> R = torch.eye(N) * r**2
+        >>> P = torch.eye(N).repeat(T, 1, 1) * p**2
+        >>> estim = torch.randn(T, N) * p
+
+        4. Perform PF prediction. Note that estimation error becomes smaller with more steps.
+
+        >>> for i in range(T - 1):
+        ...     w = q * torch.randn(N) # transition noise
+        ...     v = r * torch.randn(N) # observation noise
+        ...     states[i+1], observ[i] = model(states[i] + w, inputs[i])
+        ...     estim[i+1], P[i+1] = pf(estim[i], observ[i] + v, inputs[i], P[i], Q, R)
+        ... print('Est error:', (states - estim).norm(dim=-1))
+        Est error: tensor([5.3627, 0.5640, 0.0953, 0.0447, 0.0936])
+
+    Note:
+        Implementation is based on Section 15.2 of this book
+
+        * Dan Simon, `Optimal State Estimation: Kalman, H∞, and Nonlinear Approaches
+          <https://onlinelibrary.wiley.com/doi/epdf/10.1002/0470045345.fmatter>`_,
+          Cleveland State University, 2006
+    '''
+
+    def __init__(self, model, Q=None, R=None, particle_number=None):
+        super().__init__(model, Q, R)
+        self.particle_number = 1000 if particle_number is None else particle_number
+
+    def forward(self, x, y, u, P, Q=None, R=None, t=None, conv_weight=None):
+        r'''
+        Performs one step estimation.
+
+        Args:
+            x (:obj:`Tensor`): estimated system state of previous step
+            y (:obj:`Tensor`): system observation at current step (measurement)
+            u (:obj:`Tensor`): system input at current step
+            P (:obj:`Tensor`): state estimation covariance of previous step
+            Q (:obj:`Tensor`, optional): covariance of system transition model
+            R (:obj:`Tensor`, optional): covariance of system observation model
+
+        Return:
+            list of :obj:`Tensor`: posteriori state and covariance estimation
+        '''
+        # Upper cases are matrices, lower cases are vectors
+
+        Q = Q if Q is not None else self.Q
+        R = R if R is not None else self.R
+        conv_weight = len(x) if conv_weight is None else conv_weight
```

Comment:
this is different from doc.
Do you mean `x.size(-1)` instead of `len(x)`?
Do you mean `cov_weight` instead of `conv_weight`?
Is it better to just define `conv_weight` as `n`, which will be consistent with the doc?